🧱 Export blocks and deleting blocks

mattigrthr opened the pull request:
This PR includes a newly added export block catalog with the ability to export results as CSV files.

Also, now, blocks can be deleted from the canvas.

This PR also includes a significant state management refactoring which improved the application's overall stability.

Closes #89

arifluthfi16 committed:
📁 Export Catalog Modal & Selections (#139)

Show available export blocks in the block catalog

Commit:
🗑 Delete blocks from canvas and refactoring of state management (#140)

Delete data blocks and refactored state management

arifluthfi16 committed:
📥 CSV Export Block: Saving & Download (#142)

Download results as CSV files with an export block

Commit:
🤌🏽 Refactored tables and fixed build issue (#143)

Refactored tables and fixed build issue

floriankuwala reviewed (changes requested):
I checked out dev branch and were able to build the front end, backend and make a external database connection to florian_demo (hosted on AWS). I can see the data but the block is failed to create when configuring variables in the dataset on the canvas. I see in the cataloge section now also the export blocks however I was not able to create complete workflow because the dataset were not created on the canvas.

backend                 | ERROR:    Exception in ASGI application
backend                 | Traceback (most recent call last):
backend                 |   File "/usr/local/lib/python3.9/site-packages/uvicorn/protocols/http/httptools_impl.py", line 372, in run_asgi
backend                 |     result = await app(self.scope, self.receive, self.send)
backend                 |   File "/usr/local/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
backend                 |     return await self.app(scope, receive, send)
backend                 |   File "/usr/local/lib/python3.9/site-packages/fastapi/applications.py", line 259, in __call__
backend                 |     await super().__call__(scope, receive, send)
backend                 |   File "/usr/local/lib/python3.9/site-packages/starlette/applications.py", line 112, in __call__
backend                 |     await self.middleware_stack(scope, receive, send)
backend                 |   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 181, in __call__
backend                 |     raise exc
backend                 |   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 159, in __call__
backend                 |     await self.app(scope, receive, _send)
backend                 |   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/cors.py", line 92, in __call__
backend                 |     await self.simple_response(scope, receive, send, request_headers=headers)
backend                 |   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/cors.py", line 147, in simple_response
backend                 |     await self.app(scope, receive, send)
backend                 |   File "/usr/local/lib/python3.9/site-packages/starlette/exceptions.py", line 82, in __call__
backend                 |     raise exc
backend                 |   File "/usr/local/lib/python3.9/site-packages/starlette/exceptions.py", line 71, in __call__
backend                 |     await self.app(scope, receive, sender)
backend                 |   File "/usr/local/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
backend                 |     raise e
backend                 |   File "/usr/local/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
backend                 |     await self.app(scope, receive, send)
backend                 |   File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 656, in __call__
backend                 |     await route.handle(scope, receive, send)
backend                 |   File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 259, in handle
backend                 |     await self.app(scope, receive, send)
backend                 |   File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 61, in app
backend                 |     response = await func(request)
backend                 |   File "/usr/local/lib/python3.9/site-packages/fastapi/routing.py", line 227, in app
backend                 |     raw_response = await run_endpoint_function(
backend                 |   File "/usr/local/lib/python3.9/site-packages/fastapi/routing.py", line 162, in run_endpoint_function
backend                 |     return await run_in_threadpool(dependant.call, **values)
backend                 |   File "/usr/local/lib/python3.9/site-packages/starlette/concurrency.py", line 39, in run_in_threadpool
backend                 |     return await anyio.to_thread.run_sync(func, *args)
backend                 |   File "/usr/local/lib/python3.9/site-packages/anyio/to_thread.py", line 31, in run_sync
backend                 |     return await get_asynclib().run_sync_in_worker_thread(
backend                 |   File "/usr/local/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
backend                 |     return await future
backend                 |   File "/usr/local/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 867, in run
backend                 |     result = context.run(func, *args)
backend                 |   File "/opt/core/backend/app/routers/data_block.py", line 19, in create_data_block
backend                 |     data_block_id, model_name = data_block_controller.create_data_block(
backend                 |   File "/opt/core/backend/app/controller/data_block_controller.py", line 207, in create_data_block
backend                 |     run_dbt_models(dbt_dir=dbt_dir, dbt_model_names=dbt_model_names)
backend                 |   File "/opt/core/backend/app/controller/dbt_controller.py", line 67, in run_dbt_models
backend                 |     output.stdout.decode("utf8").split("ERROR=")[1].split("SKIP=")[0][:-1]
backend                 | IndexError: list index out of range

mattigrthr commented:
> I checked out dev branch and were able to build the front end, backend and make a external database connection to florian_demo (hosted on AWS). I can see the data but the block is failed to create when configuring variables in the dataset on the canvas. I see in the cataloge section now also the export blocks however I was not able to create complete workflow because the dataset were not created on the canvas.
> 
> backend | ERROR: Exception in ASGI application backend | Traceback (most recent call last): backend | File "/usr/local/lib/python3.9/site-packages/uvicorn/protocols/http/httptools_impl.py", line 372, in run_asgi backend | result = await app(self.scope, self.receive, self.send) backend | File "/usr/local/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in **call** backend | return await self.app(scope, receive, send) backend | File "/usr/local/lib/python3.9/site-packages/fastapi/applications.py", line 259, in **call** backend | await super().**call**(scope, receive, send) backend | File "/usr/local/lib/python3.9/site-packages/starlette/applications.py", line 112, in **call** backend | await self.middleware_stack(scope, receive, send) backend | File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 181, in **call** backend | raise exc backend | File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 159, in **call** backend | await self.app(scope, receive, _send) backend | File "/usr/local/lib/python3.9/site-packages/starlette/middleware/cors.py", line 92, in **call** backend | await self.simple_response(scope, receive, send, request_headers=headers) backend | File "/usr/local/lib/python3.9/site-packages/starlette/middleware/cors.py", line 147, in simple_response backend | await self.app(scope, receive, send) backend | File "/usr/local/lib/python3.9/site-packages/starlette/exceptions.py", line 82, in **call** backend | raise exc backend | File "/usr/local/lib/python3.9/site-packages/starlette/exceptions.py", line 71, in **call** backend | await self.app(scope, receive, sender) backend | File "/usr/local/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in **call** backend | raise e backend | File "/usr/local/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in **call** backend | await self.app(scope, receive, send) backend | File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 656, in **call** backend | await route.handle(scope, receive, send) backend | File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 259, in handle backend | await self.app(scope, receive, send) backend | File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 61, in app backend | response = await func(request) backend | File "/usr/local/lib/python3.9/site-packages/fastapi/routing.py", line 227, in app backend | raw_response = await run_endpoint_function( backend | File "/usr/local/lib/python3.9/site-packages/fastapi/routing.py", line 162, in run_endpoint_function backend | return await run_in_threadpool(dependant.call, **values) backend | File "/usr/local/lib/python3.9/site-packages/starlette/concurrency.py", line 39, in run_in_threadpool backend | return await anyio.to_thread.run_sync(func, *args) backend | File "/usr/local/lib/python3.9/site-packages/anyio/to_thread.py", line 31, in run_sync backend | return await get_asynclib().run_sync_in_worker_thread( backend | File "/usr/local/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread backend | return await future backend | File "/usr/local/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 867, in run backend | result = context.run(func, *args) backend | File "/opt/core/backend/app/routers/data_block.py", line 19, in create_data_block backend | data_block_id, model_name = data_block_controller.create_data_block( backend | File "/opt/core/backend/app/controller/data_block_controller.py", line 207, in create_data_block backend | run_dbt_models(dbt_dir=dbt_dir, dbt_model_names=dbt_model_names) backend | File "/opt/core/backend/app/controller/dbt_controller.py", line 67, in run_dbt_models backend | output.stdout.decode("utf8").split("ERROR=")[1].split("SKIP=")[0][:-1] backend | IndexError: list index out of range

That seems to be related to the demo database again. I deleted the `dbt_kuwala` schema in the database. Last time that fixed the issue. You can also delete that and look at the tables created there through tools like Datagrip or another SQL viewer.

floriankuwala commented:
Ready to Merge
